_posts/news/2024-03-19-announcing-scalajs-1.16.0.md

```diff
@@ -0,0 +1,105 @@
+
+## Release notes
+
+If upgrading from Scala.js 0.6.x, make sure to read [the release notes of Scala.js 1.0.0]({{ BASE_PATH }}/news/2020/02/25/announcing-scalajs-1.0.0/) first, as they contain a host of important information, including breaking changes.
+
+This is a **minor** release:
+
+* It is backward binary compatible with all earlier versions in the 1.x series: libraries compiled with 1.0.x through 1.15.x can be used with 1.16.0 without change.
+* It is *not* forward binary compatible with 1.15.x: libraries compiled with 1.16.0 cannot be used with 1.15.x or earlier.
+* It is *not* entirely backward source compatible: it is not guaranteed that a codebase will compile *as is* when upgrading from 1.15.x (in particular in the presence of `-Xfatal-warnings`).
+
+As a reminder, libraries compiled with 0.6.x cannot be used with Scala.js 1.x; they must be republished with 1.x first.
+
+## Enhancements with compatibility concerns
+
+### Changes to the IR and linker APIs
+
+For tooling authors who directly manipulate the IR and linker APIs, there have been some breaking changes in that area.
+This is in line with our version policy for the linker APIs.
+
+The most likely changes you may hit are:
+
+* `FieldName` is now a composite of the defining `ClassName` and a `SimpleFieldName`.
+* `StoreModule` has no parameter anymore; it implicitly applies to storing the `this` value of the enclosing module class.
+
+## Enhancements
+
+### Scala.js-specific minifier
+
+In the `fullLink` mode, the Scala.js linker now includes dedicated code size optimizations.
+They are mostly useful when the Google Closure Compiler is deactivated, which is the case by default when emitting ECMAScript modules.
+
+The most important optimization is that it compresses all *property* names (fields and methods) of Scala classes.
+It computes how frequently each one is used across the codebase, and assigns shorter names to the most frequent ones.
+
+Renaming local and global variables and functions in this way is a standard feature of general-purpose JavaScript minifiers.
+However, in general they cannot rename *properties* without breaking the semantics of JavaScript programs.
+This is why Scala.js has leveraged Closure since the very first relase.
+Closure's so-called "advanced optimizations" assume that the JavaScript program fits in a specific subset of semantics, which allows it to rename properties.
+
+Unfortunately, the way Closure handles ES modules is incompatible with what we need.
+Closure is therefore disabled when emitting ES modules from Scala.js.
+This has resulted in large bundle sizes for users of ES modules.
+
+Our minifier, which is aware of our types and of Scala.js semantics, fills that gap between Closure and general-purpose minifiers.
+When combined with a general-purpose minifier such as the one bundled with Vite, the new optimizations bring code size around 15% bigger than Closure.
```

Comment:
@sjrd @gzm0 Thanks so much for this highly awaited feature.

Just curious, what are the reasons Closure can still be better than the new Scala.js optimizer and a good general-purpose minifier? Are there hopes for a better than Closure fullOpt ES modules optimizer solution?

Reply:
Because Closure had several engineers working on it for more than a decade, probably. ;)

More precisely, because we don't do everything that could be done. At least not yet. It's likely that we will get progressively closer, over time.